### PR TITLE
Alter Migrations library

### PR DESCRIPTION
**Pull request by GregsCode12**

### CODE REVIEW

File: src/JsonConfig.cpp [hunk 1]
```diff
@@ -37,46 +38,30 @@ JsonConfig::JsonConfig(const QString &path) :
...
-    if (!file.open(QFile::ReadOnly)) {
-        return false;
-    }
-
-    QJsonDocument doc = QJsonDocument::fromJson(file.readAll());
-    QJsonValue schemaVersionsRaw = doc.object().value("schemaVersions");
-    if (!schemaVersionsRaw.isArray()) {
-        qWarning() << "Invalid schemaVersions";
+    QDir dir(m_path);
+    dir.setFilter(QDir::AllEntries | QDir::NoDotAndDotDot);
```
Comment (MasterMind2k) [inline src/JsonConfig.cpp]: It is enough only to look for text files that end with `.sql`.

File: src/JsonConfig.cpp [hunk 2]
```diff
@@ -37,46 +38,30 @@ JsonConfig::JsonConfig(const QString &path) :
...
-        }
-
-        QJsonObject schemaVersion = val.toObject();
-        QJsonValue version = schemaVersion.value("version");
-        if (!version.isDouble()) {
-            continue;
+    QStringList migrations = dir.entryList(QStringList() << "*.sql" << "*.SQL", QDir::Files);
+    foreach(QString filename, migrations)
+    {
+        QString filePath = dir.absolutePath().append("/").append(filename);
```
Comment (MasterMind2k) [inline src/JsonConfig.cpp]: Use `filePath`: https://doc.qt.io/qt-5/qdir.html#filePath

File: src/QtDBMigration.cpp [hunk 3]
```diff
@@ -32,9 +32,9 @@ class QtDBMigrationPrivate
     QtDBMigrationPrivate(QtDBMigration *q);
     ~QtDBMigrationPrivate();
 
-    int version();
```
Comment (MasterMind2k) [inline src/QtDBMigration.cpp:35]: Logic in this file needs to stay intact!

File: test/QtDBMigrationTest.cpp [hunk 4]
```diff
@@ -23,7 +23,7 @@
 #include <QString>
 #include <QtTest>
 
-#include "QtDBMigration.h"
+#include "../src/QtDBMigration.h"
```
Comment (MasterMind2k) [inline test/QtDBMigrationTest.cpp]: This means something was wrong with your build configuration

File: test/QtDBMigrationTest.pro [hunk 5]
```diff
@@ -18,13 +18,3 @@ include("../src/QtDBMigration.pri")
 
 SOURCES += \
     QtDBMigrationTest.cpp
-
-INCLUDEPATH += \
```
Comment (MasterMind2k) [inline test/QtDBMigrationTest.pro]: This should not be removed (that is why you had to fix that `#include`).

**Review by MasterMind2k — CHANGES_REQUESTED** [review]

File: src/JsonConfig.cpp (hunk 1 above)
Comment (GregsCode12) [reply src/JsonConfig.cpp]: Yeah makes sense, a more narrow filter. Changed.

File: src/JsonConfig.cpp (hunk 2 above)
Comment (GregsCode12) [reply src/JsonConfig.cpp]: Changed to use filePath.

File: src/QtDBMigration.cpp (hunk 3 above)
Comment (GregsCode12) [reply src/QtDBMigration.cpp:35]: Previous logic restored back and added the folder config.

File: test/QtDBMigrationTest.pro (hunk 5 above)
Comment (GregsCode12) [reply test/QtDBMigrationTest.pro]: Restored back to its original state.

File: test/QtDBMigrationTest.cpp (hunk 4 above)
Comment (GregsCode12) [reply test/QtDBMigrationTest.cpp]: Should be fixed.

File: CMakeLists.txt.user [hunk 6]
```diff
@@ -0,0 +1,345 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (MasterMind2k) [inline CMakeLists.txt.user]: This file should not be in repo.

Add it to .gitignore, please (separate commit).

File: src/FolderConfig.cpp [hunk 7]
```diff
@@ -0,0 +1,54 @@
+#include "FolderConfig.h"
+#include <QFile>
+#include <QDebug>
+#include<QDir>
```
Comment (MasterMind2k) [inline src/FolderConfig.cpp:4]: Need to setup automatic formatter :).

Just a remark. No need to fix.

File: src/FolderConfig.cpp [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+    QDir dir(m_path);
+    QStringList nameFilter("*.sql");
+    QStringList sqlFiles = dir.entryList(nameFilter);
+    int totalFiles = sqlFiles.count();
+
+    if (!dir.exists() || totalFiles < 1) {
+        return false;
+    }
+
+    foreach(QString filename, sqlFiles) {
```
Comment (MasterMind2k) [inline src/FolderConfig.cpp]: I think you should not use `foreach` https://www.kdab.com/goodbye-q_foreach/

You may use the C++11 `for (auto filename : sqlFiles) { ...` instead.

File: src/FolderConfig.cpp [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+        QString filePath = dir.filePath(filename);
+        QFile file(filePath);
+        if (!file.exists()) {
+            return false;
+        }
+
+        QFileInfo fileInfo(file.fileName());
+        QString fileName(fileInfo.baseName());
+        QRegExp rx("[0-9][0-9][0-9][0-9]");
```
Comment (MasterMind2k) [inline src/FolderConfig.cpp]: Why not `"\\d{4}\\.sql"`? I would like to see also file extension is validated.

File: src/FolderConfig.cpp [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+        if (!rx.exactMatch(fileName)) {
```
Comment (MasterMind2k) [inline src/FolderConfig.cpp:44]: For exact, I think you need to add `^` (start of line) and `$` (end of line) to regex.

File: src/FolderConfig.cpp [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+            continue;
+        }
+
+        if(!file.open(QFile::ReadOnly | QIODevice::Text)) {
+            return false;
+        }
+
+
+        QString fileContents = file.readAll();
```
Comment (MasterMind2k) [inline src/FolderConfig.cpp]: I think it would be better to read line per line...

Also, `\n` is only on Unix, Windows and Macs have something different.

File: src/FolderConfig.cpp [hunk 12]
```diff
@@ -0,0 +1,54 @@
...
+        QList<QString> fileSplit = fileContents.split("\n", Qt::SplitBehavior(Qt::SplitBehaviorFlags::SkipEmptyParts));
+        QString applySql = fileSplit[0];
```
Comment (MasterMind2k) [inline src/FolderConfig.cpp]: Question.

If single migration has multiple queries, will this work?

Also, one migration query can be in multiple lines.

File: src/FolderConfig.cpp [hunk 13]
```diff
@@ -0,0 +1,54 @@
...
+        QString revertSql = fileSplit[1];
+
+
+        m_schemaVersions << new SchemaVersion(
+                            file.fileName().toInt(), applySql, revertSql);
```
Comment (MasterMind2k) [inline src/FolderConfig.cpp]: Filename should have an extension.

Also, you can use `QRegExp` to extract number from whole filename.

File: src/QtDBMigration.cpp [hunk 14]
```diff
@@ -36,20 +37,27 @@ class QtDBMigrationPrivate
     bool migrate();
     bool migrate(int version);
 
+
 private:
     bool createMetaTable();
 
     QtDBMigration *m_q;
     AbstractConfig *m_config;
+    AbstractConfig *m_folder_config;
```
Comment (MasterMind2k) [inline src/QtDBMigration.cpp]: No :).

The point is, that FolderConfig is set as m_config :).

File: src/QtDBMigration.cpp [hunk 15]
```diff
@@ -36,20 +37,27 @@ class QtDBMigrationPrivate
     bool migrate();
     bool migrate(int version);
 
+
```
Comment (MasterMind2k) [inline src/QtDBMigration.cpp:40]: Why this additional empty line?

File: src/QtDBMigration.cpp [hunk 16]
```diff
@@ -36,20 +37,27 @@ class QtDBMigrationPrivate
...
     QtDBMigration *m_q;
     AbstractConfig *m_config;
+    AbstractConfig *m_folder_config;
 };
 
 QtDBMigrationPrivate::QtDBMigrationPrivate(QtDBMigration *q) :
     m_q(q) ,
-    m_config(new JsonConfig(q->m_configPath))
+    m_config(new JsonConfig(q->m_configPath)),
+    m_folder_config(new FolderConfig(q->m_configPath))
```
Comment (MasterMind2k) [inline src/QtDBMigration.cpp]: No.

From `q` should be known what type of config we want to have. We won't have a Json config and Folder config at the same time!

File: src/QtDBMigration.cpp [hunk 17]
```diff
@@ -36,20 +37,27 @@ class QtDBMigrationPrivate
...
+
 {
     // Load config
     if (!m_config->load()) {
-        qWarning() << "Fail to load migration config";
+        qWarning() << "Fail to load migration json config";
```
Comment (MasterMind2k) [inline src/QtDBMigration.cpp]: This is unneeded.

File: src/QtDBMigration.cpp [hunk 18]
```diff
@@ -84,13 +91,20 @@ int QtDBMigrationPrivate::version()
 
 bool QtDBMigrationPrivate::migrate()
 {
-    if (m_config->schemaVersions().isEmpty()) {
```
Comment (MasterMind2k) [inline src/QtDBMigration.cpp]: Do not change this. As you can see, you just duplicated code. Made the abstraction (AbstractConfig) completely useless.

File: src/QtDBMigration.cpp [hunk 19]
```diff
@@ -112,6 +126,10 @@ bool QtDBMigrationPrivate::migrate(int destVer)
                            << ", aborting...";
                 return false;
             }
+            else
+            {
+                qWarning() << "Migration successfull!! Version is:" << i;
```
Comment (MasterMind2k) [inline src/QtDBMigration.cpp]: Maybe qInfo instead? (or something similar, if none exists, use qDebug)

File: src/QtDBMigration.cpp [hunk 20]
```diff
@@ -133,6 +151,7 @@ bool QtDBMigrationPrivate::migrate(int destVer)
     return true;
 }
 
+
```
Comment (MasterMind2k) [inline src/QtDBMigration.cpp:156]: Why new empty lines?

File: src/SchemaVersion.cpp [hunk 21]
```diff
@@ -41,6 +41,7 @@ bool SchemaVersion::apply(const QSqlDatabase &db)
         upgradeVer.prepare("update qtdbmigration_meta set cur_ver = cur_ver + 1");
         return upgradeVer.exec();
     } else {
+        qDebug() << "SqLite error:" << q.lastError().text() << ", SqLite error code:" << q.lastError().number();
```
Comment (MasterMind2k) [inline src/SchemaVersion.cpp]: I would go with qWarning or even qError here.

Also note, we may use any other database engine here. So I would not use "Sqlite" term here.

File: test/QtDBMigrationTest.cpp [hunk 22]
```diff
@@ -50,7 +50,7 @@ QtDBMigrationTest::QtDBMigrationTest()
 void QtDBMigrationTest::initTestCase()
 {
     QSqlDatabase db = QSqlDatabase::addDatabase("QSQLITE");
-    db.setDatabaseName("test.db");
+    db.setDatabaseName(":memory:");
```
Comment (MasterMind2k) [inline test/QtDBMigrationTest.cpp:56]: :+1:

File: test/QtDBMigrationTest.cpp [hunk 23]
```diff
@@ -88,11 +88,15 @@ void QtDBMigrationTest::migrate_data()
     QTest::addColumn<QString>("configPath");
     QTest::addColumn<QStringList>("tableNames");
     QTest::addColumn<int>("finalVer");
-
+    QString folderPath = qEnvironmentVariable("MIGRATIONS_PATH");
     QTest::newRow("normal")
         << "test_migrate_1.json"
         << (QStringList() << "table1" << "table2")
         << 1;
+    QTest::newRow("Migrate folder: init -> 0")
```
Comment (MasterMind2k) [inline test/QtDBMigrationTest.cpp:98]: Have you also tested if migration folder is not existant?

Try to cover all edge cases (look at the `if`s and other loops in the code). Main point of tests is to cover all the lines, if possible.

Comment (MasterMind2k) [reply test/QtDBMigrationTest.cpp:98]: Have you forgot to commit the test migrations?

**Review by MasterMind2k — CHANGES_REQUESTED** [review]

File: src/FolderConfig.cpp (hunk 8 above)
Comment (GregsCode12) [reply src/FolderConfig.cpp]: Replaced.

File: src/FolderConfig.cpp (hunk 9 above)
Comment (GregsCode12) [reply src/FolderConfig.cpp]: Ok, validating with extension.

File: CMakeLists.txt.user (hunk 6 above)
Comment (GregsCode12) [reply CMakeLists.txt.user]: Added.

File: src/FolderConfig.cpp (hunk 7 above)
Comment (GregsCode12) [reply src/FolderConfig.cpp:4]: Can we try this one? :)

https://doc.qt.io/qtcreator/creator-beautifier.html

File: src/FolderConfig.cpp (hunk 10 above)
Comment (GregsCode12) [reply src/FolderConfig.cpp:44]: Yes it is true, added both symbols.

File: src/FolderConfig.cpp (hunk 11 above)
Comment (GregsCode12) [reply src/FolderConfig.cpp]: Ok, reading line per line. Split is omitted.

File: src/FolderConfig.cpp (hunk 12 above)
Comment (GregsCode12) [reply src/FolderConfig.cpp]: Should then a delimiter be used for separating up and down queries?

File: src/FolderConfig.cpp (hunk 13 above)
Comment (GregsCode12) [reply src/FolderConfig.cpp]: Changed to using regex.

File: src/QtDBMigration.cpp (hunk 14 above)
Comment (GregsCode12) [reply src/QtDBMigration.cpp]: Ok, I think it should be fixed now :)

File: src/QtDBMigration.cpp (hunk 16 above)
Comment (GregsCode12) [reply src/QtDBMigration.cpp]: Fixed, that should not be possible anymore.

File: src/QtDBMigration.cpp (hunk 17 above)
Comment (GregsCode12) [reply src/QtDBMigration.cpp]: Reverted back to a more general message.

File: src/QtDBMigration.cpp (hunk 18 above)
Comment (GregsCode12) [reply src/QtDBMigration.cpp]: Changes reverted.

File: src/QtDBMigration.cpp (hunk 19 above)
Comment (GregsCode12) [reply src/QtDBMigration.cpp]: Using qInfo for now.

File: src/SchemaVersion.cpp (hunk 21 above)
Comment (GregsCode12) [reply src/SchemaVersion.cpp]: Set to qError. Term "Sqlite" replaced with "Database engine".

File: test/QtDBMigrationTest.cpp (hunk 23 above)
Comment (GregsCode12) [reply test/QtDBMigrationTest.cpp:98]: Added more tests. Also added the test migrations. 

I pushed the new code, but will still be adding new tests so you can review as there are a few things I have to work out with Qt tests.

File: src/FolderConfig.cpp (hunk 12 above)
Comment (GregsCode12) [reply src/FolderConfig.cpp]: Ignore the question, I think it should work now :) 

I did change a little thing though. For multiple migrations, I used a list of strings, so for one query the only change is that I make a list of one query. That was also changed in JsonConfig. The rest stayed as is, hope that is ok.

File: .gitignore [hunk 24]
```diff
@@ -7,3 +7,4 @@
 *.dylib
 *.pro.user
 Makefile
+CMakeLists.txt.user
```
Comment (MasterMind2k) [inline .gitignore:10]: Heh, use full filname in commit message :).

File: migrations/0000.sql [hunk 25]
```diff
@@ -0,0 +1,5 @@
+-- +migrate Up
```
Comment (MasterMind2k) [inline migrations/0000.sql:1]: Awsome :)

File: src/FolderConfig.cpp (hunk 7 above)
Comment (MasterMind2k) [reply src/FolderConfig.cpp:4]: It was a note for future.

File: src/FolderConfig.cpp [hunk 26]
```diff
@@ -0,0 +1,94 @@
...
+#include <QFile>
+#include <QDebug>
+#include<QDir>
+
+namespace QtDBMigrationNS {
+FolderConfig::FolderConfig(const QString &path): AbstractConfig(path)
+{
+}
+
+bool FolderConfig::load()
```
Comment (MasterMind2k) [inline src/FolderConfig.cpp:23]: Can you split this function to make it more readable?

File: src/FolderConfig.cpp [hunk 27]
```diff
@@ -0,0 +1,94 @@
...
+        QRegExp rx("^\\d{4}\\.sql$");
+        if (!rx.exactMatch(fileInfo.fileName())) {
+            continue;
+        }
+
+        if(!file.open(QFile::ReadOnly | QIODevice::Text)) {
+            return false;
+        }
+
+        QTextStream in(&file);
```
Comment (MasterMind2k) [inline src/FolderConfig.cpp:51]: :+1:

File: src/FolderConfig.cpp [hunk 28]
```diff
@@ -0,0 +1,94 @@
...
+        QString line = in.readLine();
+        QStringList applySql;
+        QStringList revertSql;
+        while (!line.isNull())
```
Comment (MasterMind2k) [inline src/FolderConfig.cpp]: Try:
```cpp
do {
} while();
```

Comment (MasterMind2k) [reply src/FolderConfig.cpp]: Or, even better. Use `atEnd()`.

File: src/FolderConfig.cpp [hunk 29]
```diff
@@ -0,0 +1,94 @@
...
+        {
+            if (line.size()==0)
+            {
+                line = in.readLine();
+                continue;
+            }
+
+            if (line.contains("-- +"))
```
Comment (MasterMind2k) [inline src/FolderConfig.cpp]: Maybe use `QRegExp` instead?

Something in line `"^-- \+migrate (?P<operation>Up|Down)"`. And then you can use the captured text `operation`. Will make some ifs go away.

And also, code more readable and more robust.

File: src/FolderConfig.cpp [hunk 30]
```diff
@@ -0,0 +1,94 @@
...
+                continue;
+            }
+
+            if (migrationsUp)
+            {
+                applySql.append(line);
+            }
+            else
+            {
+                revertSql.append(line);
```
Comment (MasterMind2k) [inline src/FolderConfig.cpp]: I don't like this.

So all lines above `+migrate Up` are going to be revert migration?

File: src/FolderConfig.cpp [hunk 31]
```diff
@@ -0,0 +1,94 @@
...
+            }
+
+            line = in.readLine();
+        }
+        QRegExp rx_version("(\\d+)");
```
Comment (MasterMind2k) [inline src/FolderConfig.cpp]: What about file extension?

Do a full match. Not `indexIn`. Also file extension should match.

File: src/JsonConfig.cpp [hunk 32]
```diff
@@ -51,6 +51,9 @@ bool JsonConfig::load()
 
     QJsonArray schemaVersions = schemaVersionsRaw.toArray();
     QJsonArray::iterator it = schemaVersions.begin();
+    QStringList applySqlList;
```
Comment (MasterMind2k) [inline src/JsonConfig.cpp:54]: Why are these changes needed?

File: src/QtDBMigration.cpp [hunk 33]
```diff
@@ -43,10 +45,21 @@ class QtDBMigrationPrivate
...
+    }
+}
+
 QtDBMigrationPrivate::QtDBMigrationPrivate(QtDBMigration *q) :
-    m_q(q) ,
-    m_config(new JsonConfig(q->m_configPath))
+    m_q(q)
 {
+
+    setConfig(q->m_configPath);
```
Comment (MasterMind2k) [inline src/QtDBMigration.cpp:65]: I don't like this.

This should be explicit! If you use constructor with `QDir`, it is explicitly set that you want to have a folder config.

One better option would be to introduce static methods: 
```
fromJson(filepath)
fromDir(path)
```

This way, it is too obscure.

File: src/SchemaVersion.cpp [hunk 34]
```diff
@@ -24,8 +24,8 @@
 namespace QtDBMigrationNS {
 
 SchemaVersion::SchemaVersion(int version,
-                             const QString &applySql,
-                             const QString &revertSql) :
+                             const QStringList &applySql,
```
Comment (MasterMind2k) [inline src/SchemaVersion.cpp:27]: Why changing this?

Simply join list to QString by adding `\n`.

File: src/SchemaVersion.cpp [hunk 35]
```diff
@@ -35,27 +35,33 @@ SchemaVersion::SchemaVersion(int version,
...
-    q.prepare(m_applySql);
-    if (q.exec()) {
-        QSqlQuery upgradeVer(db);
-        upgradeVer.prepare("update qtdbmigration_meta set cur_ver = cur_ver + 1");
-        return upgradeVer.exec();
-    } else {
-        return false;
+    for (QString query : m_applySql)
+    {
+        q.prepare(query);
```
Comment (MasterMind2k) [inline src/SchemaVersion.cpp]: Ah, I see.

This won't work if several statements are inside? Good, than do this refactoring. Try to move this into separate commit (to give it why you are changing this).

Also, can you wrap all SQL statements into BEGIN/COMMIT block? It won't work on MySQL, but on other databases they will work.

File: test/QtDBMigrationTest.cpp (hunk 23 above)
Comment (MasterMind2k) [reply test/QtDBMigrationTest.cpp:98]: :+1:

File: test/QtDBMigrationTest.cpp [hunk 36]
```diff
@@ -134,6 +139,23 @@ void QtDBMigrationTest::migrateVersion_data()
...
+    bool ok = mig.migrate();
+    QVERIFY(!ok);
+}
+void QtDBMigrationTest::migrateErrors_data()
+{
+    QTest::addColumn<QString>("configPath");
+    QTest::newRow("Invalid path form")
+            << "random stirng not path";
+    QTest::newRow("Valid but empty path")
+            << QStandardPaths::writableLocation(QStandardPaths::HomeLocation);
```
Comment (MasterMind2k) [inline test/QtDBMigrationTest.cpp]: I would prefert teporary files. This way, tests are going to pollute home directory, see [this](https://doc.qt.io/qt-5/qtemporaryfile.html).

File: src/FolderConfig.cpp [hunk 37]
```diff
@@ -0,0 +1,94 @@
...
+    for(auto filename : sqlFiles) {
+        bool migrationsUp = true;
+        QString filePath = dir.filePath(filename);
+        QFile file(filePath);
+        if (!file.exists()) {
+            return false;
+        }
+
+        QFileInfo fileInfo(file.fileName());
+        QRegExp rx("^\\d{4}\\.sql$");
```
Comment (MasterMind2k) [inline src/FolderConfig.cpp]: Ah, I see you have an exact match here. Use the number out of here. No need for second regex.

File: src/FolderConfig.cpp [hunk 38]
```diff
@@ -0,0 +1,94 @@
...
+                {
+                    migrationsUp = false;
+                }
+                line = in.readLine();
+                continue;
+            }
+
+            if (migrationsUp)
+            {
+                applySql.append(line);
```
Comment (MasterMind2k) [inline src/FolderConfig.cpp]: Will this survive queries like:
```sql
alter table contacts
  modify col1 int,
  modify col2 varchar,
  add col2 int;
```

Try adding a test first, and fix later ;).

**Review by MasterMind2k — CHANGES_REQUESTED** [review]
Good progress!

Keep up the good work.

File: .gitignore (hunk 24 above)
Comment (GregsCode12) [reply .gitignore:10]: Fixed.

File: src/FolderConfig.cpp (hunk 26 above)
Comment (GregsCode12) [reply src/FolderConfig.cpp:23]: Function split.

File: test/QtDBMigrationTest.cpp (hunk 36 above)
Comment (GregsCode12) [reply test/QtDBMigrationTest.cpp]: Changed to temp files.

File: src/FolderConfig.cpp (hunk 28 above)
Comment (GregsCode12) [reply src/FolderConfig.cpp]: Done.

File: src/FolderConfig.cpp (hunk 29 above)
Comment (GregsCode12) [reply src/FolderConfig.cpp]: Changed to using ```QRegularExpression```.

File: src/FolderConfig.cpp (hunk 30 above)
Comment (GregsCode12) [reply src/FolderConfig.cpp]: Added ```else if``` for Down migrations and an ```else``` in case of random text outside of migrations that gets ignored.

File: src/FolderConfig.cpp (hunk 31 above)
Comment (GregsCode12) [reply src/FolderConfig.cpp]: Fixed this part according to a comment below.

File: src/FolderConfig.cpp (hunk 37 above)
Comment (GregsCode12) [reply src/FolderConfig.cpp]: Changed to use number from the first RegEx.

File: src/SchemaVersion.cpp (hunk 35 above)
Comment (GregsCode12) [reply src/SchemaVersion.cpp]: Added to a separate commit. SQL statements wrapped in transaction block.

File: src/FolderConfig.cpp (hunk 38 above)
Comment (GregsCode12) [reply src/FolderConfig.cpp]: Nevermind, I see what you mean. Adding fix.

File: src/FolderConfig.cpp [hunk 39]
```diff
@@ -0,0 +1,92 @@
...
+
+    for(auto fileName : sqlFiles) {
+        QString operation = "";
+        QString filePath = dir.filePath(fileName);
+        QFile file(filePath);
+        if(!file.exists())
+        {
+            return false;
+        }
+        QFileInfo fileInfo(file.fileName());
```
Comment (MasterMind2k) [inline src/FolderConfig.cpp]: Why do you need `QFileInfo`?

File: src/QtDBMigration.cpp [hunk 40]
```diff
@@ -43,10 +45,24 @@ class QtDBMigrationPrivate
     AbstractConfig *m_config;
 };
 
+void QtDBMigrationPrivate::setConfig(QString path) {
+    if (QFileInfo::exists(path))
```
Comment (MasterMind2k) [inline src/QtDBMigration.cpp:49]: Are you sure this is correct?

Have you tested the current functionality still works?

File: src/QtDBMigration.cpp [hunk 41]
```diff
@@ -43,10 +45,24 @@ class QtDBMigrationPrivate
...
+    {
+        QFileInfo inputPath(path);
+        if (inputPath.isDir() && inputPath.exists()) {
+        m_config = new FolderConfig(path);
+        }
```
Comment (MasterMind2k) [inline src/QtDBMigration.cpp:54]: Why not putting else clause here?

File: src/QtDBMigration.cpp (hunk 40 above)
Comment (MasterMind2k) [reply src/QtDBMigration.cpp:49]: I see, it checks if *file* exists.

Ok.

**Review by MasterMind2k — CHANGES_REQUESTED** [review]
One small thing to look at

File: src/QtDBMigration.cpp (hunk 41 above)
Comment (MasterMind2k) [reply src/QtDBMigration.cpp:54]: Irelevant.

File: src/QtDBMigration.cpp (hunk 40 above)
Comment (MasterMind2k) [reply src/QtDBMigration.cpp:49]: Irelevant.

File: src/FolderConfig.cpp (hunk 39 above)
Comment (GregsCode12) [reply src/FolderConfig.cpp]: Used it more at first, now I see it is only in one statement and not really needed. Removed it.

**Review by MasterMind2k — APPROVED** [review]
